File: 2023/02/aoc02-map-nums.rb
#!/usr/bin/env ruby

class MappingEntry
  attr_reader :value, :text
  def initialize(value, text)
    @value = value
    @text = text
  end
end

token_mappings = Array.new

token_mappings.push(MappingEntry.new(1, 'one'))
token_mappings.push(MappingEntry.new(2, 'two'))
token_mappings.push(MappingEntry.new(3, 'three'))
token_mappings.push(MappingEntry.new(4, 'four'))
token_mappings.push(MappingEntry.new(5, 'five'))
token_mappings.push(MappingEntry.new(6, 'six'))
token_mappings.push(MappingEntry.new(7, 'seven'))
token_mappings.push(MappingEntry.new(8, 'eight'))
token_mappings.push(MappingEntry.new(9, 'nine'))

# TODO boundary checking
def is_match?(candidate, mapping, position)

  if candidate.length < position - 1 || mapping.text.length < position - 1
    puts "position is greater than candidate length"
    return
  end

  if candidate[position] == mapping.text[position]
    if position == mapping.text.length - 1
      return true
    else
      return is_match?(candidate, mapping, position+1)
    end
  end

  false
end

def numeric? string
  true if Float(string) rescue false
end

def process_line(line, mappings)
  new_line = ''
  next_position = 0
  match_found = false

  last_match_position = 0

  (0..line.length-1).each do | position|
    if next_position > position
      next
    end
    mappings.each do |mapping|
      if is_match?(line[position..-1], mapping, 0)
        new_line << mapping.value.to_s
        last_match_position = position + mapping.text.length
        next_position = position + mapping.text.length - 1
        match_found = true
        break
      else
        if match_found
          match_found = false
        end
      end
    end

    unless match_found
      if last_match_position <= position
        last_match_position = position
        new_line << line[position, 1]
      end
      match_found = false
    end
  end

  new_line
end

ARGF.each_line do |line|
  print process_line(line, token_mappings)
end

File: 2023/01/part2/aoc02-map-nums.pl
#!/usr/bin/env perl

use warnings;
use strict;

use Scalar::Util qw(looks_like_number);
use Data::Dumper;


## TODO
# - everything should be lowercase

while(<>){
    
    my %mappings = ( one => 1,
		 two => 2,
		 three => 3,
		 four => 4,
		 five => 5,
		 six => 6,
		 seven => 7,
		 eight => 8,
		     nine => 9 );

    process_line(lc($_), \%mappings);
}


sub process_line {
    my $line = shift;
    my $mappingsRef = shift;

    my $result = "";
    my $token = "";

    foreach my $char (split //, $line){
	if(looks_like_number($char)){
	    $result = $result . $token . $char;
	    $token = "";
	}
	else {   
	    $token = $token . $char;
	    $token = getMappedToken($token, $mappingsRef);
	}	    
    }
    $result = $result . $token;

    printf("%s", $result);
}

sub getMappedToken(){
    my $token = shift;
    my $mappingsRef = shift;
    
    foreach my $key (keys %{$mappingsRef}) {
	if(index($token, $key) != -1){
	    my $mappedValue = $mappingsRef->{$key};
	    my $newToken = $token =~ s/$key/$mappedValue/r;
	    return $newToken;
	}
    }

    return $token;
}
